.travis.yml

```diff
@@ -36,6 +36,25 @@ jobs:
           after_script:
               - $CI_MANAGERS/fedora.sh CLEANUP
 
+        - stage: Build & test with ASan
```

Comment:
In my opinion it would be great if both standard and ASAN runs run in parallel - dropping the stage definition here should do the trick.

Reply:
I tried something like that but it didn't work out. Travis just ran the ASan part skipping the "basic" part altogether. Maybe, I got the indentation wrong. I'll try again to see how it goes.

Reply:
Just to clarify what I meant by "later". My idea was to make Travis work and merge it as soon as possible so that the CI could start catching issues that are easily detectable by ASan+UBsan and the stages/jobs could be rearranged later in a way that wouldn't require catching up with new bugs lurking around :-)

Reply:
Here's a working *fix* for the parallel runs (e.g. https://travis-ci.org/mrc0mmand/systemd/jobs/452344355)

```patch
From a084480b15b817b95f5db4e1605d88635c1fdd99 Mon Sep 17 00:00:00 2001
From: Frantisek Sumsal <frantisek@sumsal.cz>
Date: Thu, 8 Nov 2018 12:47:07 +0100
Subject: [PATCH] travis: run ASan job in parallel with the standard one

---
 .travis.yml | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/.travis.yml b/.travis.yml
index add0576d4..eaeb95051 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -36,8 +36,7 @@ jobs:
           after_script:
               - $CI_MANAGERS/fedora.sh CLEANUP
 
-        - stage: Build & test with ASan
-          name: Fedora Rawhide
+        - name: Fedora Rawhide (ASan)
           language: bash
           env:
               - FEDORA_RELEASE="rawhide"
-- 
2.17.2
```

Reply:
Thank you! Now I'm a bit puzzled about what went wrong when I tried to do the same thing :-) Anyway, I applied the patch. Let's see how it goes.

Reply:
Well, in *my* run there's a timeout in `test-execute`: https://travis-ci.org/mrc0mmand/systemd/jobs/452344355. Did it happen in previous ASan/UBSan runs as well?

Reply:
`test-execute` had been failing until https://github.com/systemd/systemd/issues/10677 was merged. The last time the test were run it passed but `test-capability` failed with:
```
--- command ---
SYSTEMD_KBD_MODEL_MAP='/build/src/locale/kbd-model-map' PATH='/build/build:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin' SYSTEMD_LANGUAGE_FALLBACK_MAP='/build/src/locale/language-fallback-map' /build/build/test-capability
--- stderr ---
have ambient caps: yes
Capabilities:= cap_chown,cap_dac_override,cap_dac_read_search,cap_fowner,cap_fsetid,cap_kill,cap_setgid,cap_setuid,cap_setpcap,cap_linux_immutable,cap_net_bind_service,cap_net_broadcast,cap_net_admin,cap_net_raw,cap_ipc_lock,cap_ipc_owner,cap_sys_module,cap_sys_rawio,cap_sys_chroot,cap_sys_ptrace,cap_sys_pacct,cap_sys_admin,cap_sys_boot,cap_sys_nice,cap_sys_resource,cap_sys_time,cap_sys_tty_config,cap_mknod,cap_lease,cap_audit_write,cap_audit_control,cap_setfcap,cap_mac_override,cap_mac_admin,cap_syslog,cap_wake_alarm,cap_block_suspend,cap_audit_read+eip
Capabilities:= cap_dac_override,cap_net_raw+ep
==7021==LeakSanitizer has encountered a fatal error.
==7021==HINT: For debugging, try setting environment variable LSAN_OPTIONS=verbosity=1:log_threads=1
==7021==HINT: LeakSanitizer does not work under ptrace (strace, gdb, etc)
Assertion 'WIFEXITED(status) && WEXITSTATUS(status) == 0' failed at ../src/test/test-capability.c:71, function fork_test(). Aborting.
-------
```
but generally, yes, `test-execute` seems to be flaky and probably the timeout should be increased a bit.

Reply:
I see, thanks for the info. I'm looking into the `test-capability` fail right now.

Reply:
In the meantime, I'm going to put the kludge there again and use `meson test -t 3` to run the tests with 3 times the normal timeout. The PR is already useful as it is even without `test-capability`, which can be fixed later.

Reply:
Ugh, this is, once again, some black magic stuff.

Let's take a look at a more verbose version of the log:
```
==5649==AddressSanitizer Init done
have ambient caps: yes
Capabilities:= cap_chown,cap_dac_override,cap_dac_read_search,cap_fowner,cap_fsetid,cap_kill,cap_setgid,cap_setuid,cap_setpcap,cap_linux_immutable,cap_net_bind_service,cap_net_broadcast,cap_net_admin,cap_net_raw,cap_ipc_lock,cap_ipc_owner,cap_sys_module,cap_sys_rawio,cap_sys_chroot,cap_sys_ptrace,cap_sys_pacct,cap_sys_admin,cap_sys_boot,cap_sys_nice,cap_sys_resource,cap_sys_time,cap_sys_tty_config,cap_mknod,cap_lease,cap_audit_write,cap_audit_control,cap_setfcap,cap_mac_override,cap_mac_admin,cap_syslog,cap_wake_alarm,cap_block_suspend,cap_audit_read+eip
Capabilities:= cap_dac_override,cap_net_raw+ep
==5656==Could not attach to thread 5655 (errno 1).
==5656==Failed suspending threads.
==5655==LeakSanitizer has encountered a fatal error.
==5655==HINT: For debugging, try setting environment variable LSAN_OPTIONS=verbosity=1:log_threads=1
==5655==HINT: LeakSanitizer does not work under ptrace (strace, gdb, etc)
```

From the issues and documentation I found so far it means that Docker is missing the `SYS_PTRACE` capability. However, as shown in the output above, the capability is there (and `docker run --cap-add CAP_PTRACE` makes no difference).

And to make matters worse, it, of course, works as expected on both CentOS and Fedora...

Reply:
Okay, after another painfully long debug session I finally found a *fix* (or workaround, still not sure). The fail is caused by `test_drop_privileges_fail`, as it drops all privileges, even `SYS_PTRACE`, which is necessary for ASan. I have no idea why it works on Fedora/CentOS, but the proposed patch below seems to fix the issue:

```patch
diff --git a/src/test/test-capability.c b/src/test/test-capability.c
index 730dbf8cb..fc0d347bb 100644
--- a/src/test/test-capability.c
+++ b/src/test/test-capability.c
@@ -21,7 +21,7 @@ static uid_t test_uid = -1;
 static gid_t test_gid = -1;
 
 /* We keep CAP_DAC_OVERRIDE to avoid errors with gcov when doing test coverage */
-static uint64_t test_flags = 1ULL << CAP_DAC_OVERRIDE;
+static uint64_t test_flags = (1ULL << CAP_DAC_OVERRIDE) | (1ULL << CAP_SYS_PTRACE);
```

@evverx - If it's a relatively sane solution, I'll open another PR.

Reply:
@mrc0mmand I've just opened https://github.com/systemd/systemd/issues/10692. Let's continue the discussion there.